modules/map/tools/map_datachecker/server/eight_route.cc

```diff
@@ -0,0 +1,172 @@
+      || _sp_good_pose_info->start_time < 0
+      || _sp_good_pose_info->end_time < 0) {
+    return 0.0;
+  }
+  return _sp_good_pose_info->end_time - _sp_good_pose_info->start_time;
+}
+
+double EightRoute::get_eight_route_progress(
+  const std::vector<FramePose> & poses) {
+  int size = static_cast<int>(poses.size());
+  int start_index = time_to_index(poses, _start_time);
+  // select first good pose
+  while (start_index < size) {
+    if (is_good_pose(poses, start_index)
+        && is_eight_route_pose(poses, start_index)) {
+      AINFO << "find first good pose.index:" << start_index;
+      break;
+    }
+    ++start_index;
+  }
+  if (start_index >= size) {
+    AINFO << "not find first good pose, start_time: "
+          << std::to_string(_start_time)
+          << ", start_index: "<< start_index
+          << ", pose size: " << size;
+    return 0.0;
+  }
+  if (start_index + 1 >= size) {
+    AINFO << "not have enough poses, wait for a moment";
+    return 0.0;
+  }
+  _last_yaw = get_yaw(poses[start_index].tx,
+    poses[start_index].ty,
+    poses[start_index+1].tx,
+    poses[start_index+1].ty);
+
+  int not_eight_count = 0;
+  for (int i = start_index + 2; i < size; ++i) {
+    if (!is_good_pose(poses, i)) {
+      AINFO << "not good pose";
+      return 0.0;
+    }
+    if (!is_eight_route_pose(poses, i)) {
+      ++not_eight_count;
+      AINFO << "not eight route pose";
+      if (not_eight_count > _sp_conf->eight_bad_pose_tolerance) {
+        AINFO << "not-eight pose count reached upper limitation";
+        _return_state = ErrorCode::ERROR_NOT_EIGHT_ROUTE;
+        return 0.0;
+      }
+    } else {
+      not_eight_count = 0;
+    }
+    AINFO << "good pose";
+    update_good_pose_info(poses[i]);
+  //  clear_bad_pose_info();
+  }
+  double eight_route_during = get_good_pose_during();
+  if (eight_route_during < 1e-8) {
+    AINFO << "num of eight route good pose too small, during: "
+        << eight_route_during;
+    _return_state = ErrorCode::SUCCESS;
+    return 0.0;
+  }
+  _return_state = ErrorCode::SUCCESS;
+  double progress = eight_route_during / _sp_conf->eight_duration;
+  if (progress >= 1.0) {
+    progress = 1.0;
+  }
+  clear_good_pose_info();
+  return progress;
+}
+
+ErrorCode EightRoute::process(const std::vector<FramePose>& poses) {
+  AINFO << "[EightRoute::process] begin";
+  size_t size = poses.size();
+  if (size <= 1) {
+    _return_state = ErrorCode::ERROR_VERIFY_NO_GNSSPOS;
+    return _return_state;
+  }
+
+  _progress = get_eight_route_progress(poses);
+  if (_return_state != ErrorCode::SUCCESS) {
+    AINFO << "get_eight_route_progress failed.";
+    return _return_state;
+  }
+  if (_progress < _last_progress) {
+    _return_state = ErrorCode::ERROR_NOT_EIGHT_ROUTE;
+    return _return_state;
+  }
+
+  AINFO << "[EightRoute::process] end, progress:" << _progress;
+  _return_state = ErrorCode::SUCCESS;
+  return _return_state;
+}
+
+double EightRoute::get_progress() {
```

Comment:
add const to the end of class for getter.

Reply:
OK